src/test_memcachedsolution.cpp

```diff
@@ -0,0 +1,864 @@
+
+  TypeParam::fail_memcached_instance(this->_memcached_instances.back());
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::OK, rc);
+  EXPECT_EQ(data_out, data_in);
+
+  TypeParam::restore_memcached_instance(this->_memcached_instances.back());
+}
+
+/// Add a key. Kill a memcached instance. Try retrieve the key after it should
+/// have expired.
+TYPED_TEST(SimpleMemcachedSolutionTest, AddKillGetExpire)
+{
+  uint64_t cas = 0;
+  Store::Status rc;
+  std::string data_in = "SimpleMemcachedSolutionTest.AddKillGetExpire";
+  std::string data_out;
+
+  rc = this->set_data(data_in, cas, 1);
+  EXPECT_EQ(Store::Status::OK, rc);
+
+  TypeParam::fail_memcached_instance(this->_memcached_instances.back());
+
+  sleep(2);
+
+  rc = this->get_data(data_out, cas);
+  EXPECT_EQ(Store::Status::NOT_FOUND, rc);
+
+  TypeParam::restore_memcached_instance(this->_memcached_instances.back());
+}
+
+/// Add a key and retrieve it. Kill a memcached instance. Update the key. This
+/// sometimes result in data contention depending on whether the primary or
+/// backup memcached for this key has been killed. If it does, retrieve the key
+/// again and update the key. This will work. If the first update worked, do
+/// another update and check this fails due to data contention.
+///
+/// Repeat this test 10 times so that we sometimes kill the primary and
+/// sometimes the backup.
+TYPED_TEST(SimpleMemcachedSolutionTest, AddKillSetSetDataContentionSet)
+{
+  for (int ii = 0; ii < 10; ++ii)
+  {
+    this->get_new_key();
+
+    uint64_t cas = 0;
+    Store::Status rc;
+    std::string data_in = "SimpleMemcachedSolutionTest.AddKillSetSetDataContentionSet";
+    std::string data_out;
+
+    rc = this->set_data(data_in, cas);
+    EXPECT_EQ(Store::Status::OK, rc);
+
+    rc = this->get_data(data_out, cas);
+    EXPECT_EQ(Store::Status::OK, rc);
+    EXPECT_EQ(data_out, data_in);
+
+    TypeParam::fail_memcached_instance(this->_memcached_instances.back());
+
+    data_in = "SimpleMemcachedSolutionTest.AddKillSetSetDataContentionSet_New1";
+    rc = this->set_data(data_in, cas);
+
+    if (rc == Store::Status::DATA_CONTENTION)
+    {
+      rc = this->get_data(data_out, cas);
+      EXPECT_EQ(Store::Status::OK, rc);
+
+      rc = this->set_data(data_in, cas, 0);
+      EXPECT_EQ(Store::Status::OK, rc);
+    }
+    else if (rc == Store::Status::OK)
+    {
+      std::string failed_data_in = "FAIL";
+      rc = this->set_data(data_in, cas);
+      EXPECT_EQ(Store::Status::DATA_CONTENTION, rc);
+
+      rc = this->get_data(data_out, cas);
+      EXPECT_EQ(Store::Status::OK, rc);
+      EXPECT_EQ(data_out, data_in);
+
+      data_in = "SimpleMemcachedSolutionTest.AddKillSetSetDataContentionSet_New2";
+      rc = this->set_data(data_in, cas);
+      EXPECT_EQ(Store::Status::OK, rc);
+
+      rc = this->get_data(data_out, cas);
+      EXPECT_EQ(Store::Status::OK, rc);
+      EXPECT_EQ(data_out, data_in);
+    }
+    else
+    {
+      EXPECT_TRUE(false);
```

Comment:
Presumably this doesn't give a very nice error message if we ever hit this branch. An `EXPECT_TRUE` on `rc` at the start of the branch might be nicer?